[motion-path] Update "<basic-shape>" implementation for offset-path

nmoucht opened the pull request:
#### 5c26aeef463837637ee7d33492845821ae276e25
<pre>
[motion-path] Update &quot;&lt;basic-shape&gt;&quot; implementation for offset-path
<a href="https://bugs.webkit.org/show_bug.cgi?id=259994">https://bugs.webkit.org/show_bug.cgi?id=259994</a>
rdar://110565070

Reviewed by Tim Nguyen.

Update ShapePathOperation to use the containing block&apos;s rect.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-shape-circle-001.html:
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setContainingBlockReferenceRect):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setReferenceBoxForPathOperations):

Canonical link: <a href="https://commits.webkit.org/267556@main">https://commits.webkit.org/267556@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/029cc262a32d1164ace485cedd980662c224b195

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17671 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15812 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17352 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18071 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17090 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17449 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19481 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14692 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22050 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15676 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19802 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13636 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15255 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15156 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19618 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2086 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/MotionPath.cpp:
```diff
@@ -62,6 +62,15 @@ std::optional<MotionPathData> MotionPath::motionPathDataForRenderer(const Render
         }
         return std::nullopt;
     }
+    if (is<ShapePathOperation>(pathOperation)) {
+        if (auto* container = renderer.containingBlock()) {
+            auto& shapePathOperation = downcast<ShapePathOperation>(*pathOperation);
+            data.containingBlockBoundingRect = snapRectToDevicePixelsIfNeeded(container->referenceBoxRect(shapePathOperation.referenceBox()), downcast<RenderLayerModelObject>(renderer));
+            data.offsetFromContainingBlock = offsetFromContainer(renderer, *container);
+            return data;
+        }
+        return std::nullopt;
+    }
```
I wonder if we can cleanly share code somehow with `BoxPathOperation`. Not a big deal if we can't though.

nmoucht replied:
Yeah I'm planning on sharing them once referenceBox() goes on the root PathOperation in https://github.com/WebKit/WebKit/pull/17140.

nt1m reviewed (approved):
Looks like iOS might need some fuzziness adjustments.

(Also can we export those to WPT? `export-w3c-test-changes -g HEAD -b 259994 -c --no-linter`)

nmoucht committed:
[motion-path] Update "<basic-shape>" implementation for offset-path
https://bugs.webkit.org/show_bug.cgi?id=259994
rdar://110565070

Reviewed by Tim Nguyen.

Update ShapePathOperation to use the containing block's rect.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/offset-path-shape-circle-001.html:
* Source/WebCore/rendering/PathOperation.h:
(WebCore::PathOperation::setContainingBlockReferenceRect):
(WebCore::PathOperation::PathOperation):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setReferenceBoxForPathOperations):

Canonical link: https://commits.webkit.org/267556@main
